src/Data/Text/IO.hs

```diff
@@ -176,17 +179,24 @@ hGetLine = hGetLineWith T.concat
 hPutStr :: Handle -> Text -> IO ()
 -- This function is lifted almost verbatim from GHC.IO.Handle.Text.
 hPutStr h t = do
-  (buffer_mode, nl) <-
+  (buffer_mode, nl, isUTF8) <-
        wantWritableHandle "hPutStr" h $ \h_ -> do
                      bmode <- getSpareBuffer h_
-                     return (bmode, haOutputNL h_)
-  let str = stream t
+                     let isUTF8 = maybe False ((== "UTF-8") . textEncodingName) $ haCodec h_
+                     return (bmode, haOutputNL h_, isUTF8)
+  let l = T.length t
+      str = stream t
   case buffer_mode of
-     (NoBuffering, _)        -> hPutChars h str
-     (LineBuffering, buf)    -> writeLines h nl buf str
-     (BlockBuffering _, buf)
-         | nl == CRLF        -> writeBlocksCRLF h buf str
-         | otherwise         -> writeBlocksRaw h buf str
+    (NoBuffering, _)
+      | l > 5 && nl == LF && isUTF8   -> utf8Put
+      | otherwise                     -> hPutChars h str
+    _ | l > 80 && nl == LF && isUTF8  -> utf8Put
```

Comment:
Why only if `l > 80`, or `l > 5` in the `NoBuffering` case?

Reply:
Based on #585 benchmarks

Reply:
Those magic numbers seem too fragile. They are heavily dependent on OS, hardware, and the auxiliary functions that do the actual writing. In any case I wouldn't consider `hPutStr` a performance-critical function that makes this kind of trick worth the loss in maintainability.

Branch on the simple condition without length. `utf8Put` can be optimized later if someone cares (a low hanging fruit is to avoid the copy in `encodeUtf8`).